tests/lib/openstack_query/test_query_parser.py

```diff
@@ -0,0 +1,415 @@
+        Should call run_sort method, and return result
+        """
+        self.instance._sort_by = (MockProperties.PROP_1, False)
+        mock_run_sort.return_value = "sort-out"
+        mock_obj_list = ["obj1", "obj2", "obj3"]
+        res = self.instance.run_parser(mock_obj_list)
+
+        self.assertEqual(res, "sort-out")
+        mock_run_sort.assert_called_once_with(mock_obj_list, self.instance._sort_by)
+
+    @patch("openstack_query.query_parser.QueryParser._run_group_by")
+    def test_run_parser_with_group_mappings(self, mock_run_group_by):
+        """
+        Tests that run_parser functions expectedly - when giving group_mappings
+        Should call run_group_by with group mappings and return
+        """
+        self.instance._group_by = MockProperties.PROP_1
+        self.instance._group_mappings = {"group1": "some-mapping_func"}
+        mock_obj_list = ["obj1", "obj2", "obj3"]
+        mock_run_group_by.return_value = "group-out"
+        self.instance._group_by = MockProperties.PROP_1
+
+        res = self.instance.run_parser(mock_obj_list)
+        self.assertEqual(res, "group-out")
+        mock_run_group_by.assert_called_once_with(
+            mock_obj_list, self.instance._group_mappings
+        )
+
+    @patch("openstack_query.query_parser.QueryParser._build_unique_val_groups")
+    @patch("openstack_query.query_parser.QueryParser._run_group_by")
+    def test_run_parser_with_no_group_mappings(
+        self, mock_run_group_by, mock_build_unique_val_groups
+    ):
+        """
+        Tests that run_parser functions expectedly - when giving group_by with no group_mappings
+        Should call build_unique_val_groups first to set group_mappings and then call run_group_by and return
+        """
+        self.instance._group_by = MockProperties.PROP_1
+        mock_obj_list = ["obj1", "obj2", "obj3"]
+        mock_build_unique_val_groups.return_value = "group-mapping-out"
+        mock_run_group_by.return_value = "group-out"
+
+        res = self.instance.run_parser(mock_obj_list)
+        self.assertEqual(res, "group-out")
+        mock_build_unique_val_groups.assert_called_once_with(
+            mock_obj_list, group_by_prop=MockProperties.PROP_1
+        )
+        mock_run_group_by.assert_called_once_with(mock_obj_list, "group-mapping-out")
+
+    @patch("openstack_query.query_parser.QueryParser._run_sort")
+    @patch("openstack_query.query_parser.QueryParser._run_group_by")
+    def test_run_parser_with_sort_and_group(self, mock_run_group_by, mock_run_sort):
+        """
+        Tests that run_parser functions expectedly - when giving both group_by and sort_by
+        Should call run_sort method and then run_group_by on that output, then return
+        """
+        self.instance._sort_by = (MockProperties.PROP_1, False)
+        self.instance._group_by = MockProperties.PROP_1
+        self.instance._group_mappings = {"group1": "some-mapping_func"}
```

Comment:
Ditto about sort_by and group_by having getters

For `group_mappings" we should call the public method `setup_groupings` instead of circumventing it. If it's a hassle to do this it either means our API needs updating or we need a helper method in the test or a fake object....etc.

I'm sure I've missed other places we do this, but it's more me trying to get us out of the habit of using protected by default and "dog food" the interface we're providing from objects in tests

Reply:
It's basically this in a nutshell:
https://softwareengineering.stackexchange.com/a/380296

It's not going to cause problems, but it's effectively us saying we don't need to consider how this object works from a public API point of view or the fact it's large enough that we can't observe it through a public method

Reply:
I'll split this out into a separate PR - since we do this in lots of places #108